engine/internal/runtime/manager.go

```diff
@@ -0,0 +1,214 @@
+func (m *Manager) ListSyncedModelIDs(ctx context.Context) []string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	var modelIDs []string
+	for id := range m.readyRuntimes {
+		modelIDs = append(modelIDs, id)
+	}
+	return modelIDs
+}
+
+// ListInProgressModels returns the list of models that are in progress.
+func (m *Manager) ListInProgressModels() []string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	var modelIDs []string
+	for id := range m.pendingRuntimes {
+		modelIDs = append(modelIDs, id)
+	}
+	return modelIDs
+}
+
+// PullModel pulls the model from the model manager.
+func (m *Manager) PullModel(ctx context.Context, modelID string) error {
+	if m.isReady(modelID) {
+		return nil
+	}
+
+	var done chan struct{}
+	m.mu.Lock()
+	if ch, ok := m.pendingRuntimes[modelID]; ok {
+		m.mu.Unlock()
+		done = ch
+
+	} else {
+		done = make(chan struct{})
+		m.pendingRuntimes[modelID] = done
+		m.mu.Unlock()
+
+		if err := m.rtClient.DeployRuntime(ctx, modelID); err != nil {
+			m.mu.Lock()
+			delete(m.pendingRuntimes, modelID)
+			m.mu.Unlock()
+			return err
+		}
+	}
+
+	ctrl.LoggerFrom(ctx).Info("Waiting for runtime to be ready", "model", modelID)
+	select {
+	case <-done:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return nil
+}
+
+// Reconcile reconciles the runtime.
+func (m *Manager) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	var sts appsv1.StatefulSet
+	if err := m.k8sClient.Get(ctx, req.NamespacedName, &sts); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	modelID := sts.GetAnnotations()[modelAnnotationKey]
+
+	if !sts.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&sts, finalizerKey) {
+			return ctrl.Result{}, nil
+		}
+		m.deleteRuntime(modelID)
+		controllerutil.RemoveFinalizer(&sts, finalizerKey)
+		if err := m.k8sClient.Update(ctx, &sts); err != nil {
+			log.Error(err, "Failed to remove finalizer")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if sts.Status.ReadyReplicas > 0 && m.isPending(modelID) {
+		m.markRuntimeReady(modelID, m.rtClient.GetAddress(sts.Name))
+		log.Info("Runtime is ready")
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the runtime manager with the given controller manager.
+func (m *Manager) SetupWithManager(mgr ctrl.Manager) error {
+	filterByAnno := (predicate.NewPredicateFuncs(func(object client.Object) bool {
+		_, ok := object.GetAnnotations()[runtimeAnnotationKey]
+		return ok
+	}))
+	constructer := func(r *reconcile.Request) logr.Logger {
+		if r != nil {
+			return mgr.GetLogger().WithValues("runtime", r.NamespacedName)
+		}
+		return mgr.GetLogger()
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.StatefulSet{}, builder.WithPredicates(filterByAnno)).
+		WithLogConstructor(constructer).
+		Complete(m)
+}
+
+// PreloadRuntimes preloads the runtimes. This function is not thread-safe.
```

Comment:
How about initialize(Manager)? Initially, I did this process inside the `NewManager()`, but since preloading requires different arguments, I separated the logic into this function.

Reply:
sounds good!